sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java

```diff
@@ -0,0 +1,317 @@
+        List<Integer> orderKeys,
+        List<Boolean> orderOrientations,
+        List<Boolean> orderNulls,
+        int lowerLimit,
+        int upperLimit,
+        boolean rows,
+        List<Integer> inputFields,
+        Combine.CombineFn combineFn,
+        Schema.Field outputField) {
+      this.partitionKeys = partitionKeys;
+      this.orderKeys = orderKeys;
+      this.orderOrientations = orderOrientations;
+      this.orderNulls = orderNulls;
+      this.lowerLimit = lowerLimit;
+      this.upperLimit = upperLimit;
+      this.rows = rows;
+      this.inputFields = inputFields;
+      this.combineFn = combineFn;
+      this.outputField = outputField;
+    }
+  }
+
+  @Override
+  public NodeStats estimateNodeStats(RelMetadataQuery mq) {
+    NodeStats inputStat = BeamSqlRelUtils.getNodeStats(this.input, mq);
+    return inputStat;
+  }
+
+  @Override
+  public BeamCostModel beamComputeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+    NodeStats inputStat = BeamSqlRelUtils.getNodeStats(this.input, mq);
+    float multiplier = 1f + 0.125f;
+    return BeamCostModel.FACTORY.makeCost(
+        inputStat.getRowCount() * multiplier, inputStat.getRate() * multiplier);
+  }
+
+  private static class Transform extends PTransform<PCollectionList<Row>, PCollection<Row>> {
+
+    private Schema outputSchema;
+    private List<FieldAggregation> aggFields;
+
+    public Transform(Schema schema, List<FieldAggregation> fieldAgg) {
+      this.outputSchema = schema;
+      this.aggFields = fieldAgg;
+    }
+
+    @Override
+    public PCollection<Row> expand(PCollectionList<Row> input) {
+      PCollection<Row> inputData = input.get(0);
+      Schema inputSchema = inputData.getSchema();
+      for (FieldAggregation af : aggFields) {
+        if (af.partitionKeys.isEmpty()) {
+          // This sections simulate a KV Row
+          // Similar to the output of Group.byFieldIds
+          // When no partitions are specified
+          Schema inputSch = inputData.getSchema();
+          Schema mockKeySchema =
+              Schema.of(Schema.Field.of("mock", Schema.FieldType.STRING.withNullable(true)));
+          Schema simulatedKeyValueSchema =
+              Schema.of(
+                  Schema.Field.of("key", Schema.FieldType.row(mockKeySchema)),
+                  Schema.Field.of(
+                      "value", Schema.FieldType.iterable(Schema.FieldType.row(inputSch))));
+          PCollection<Iterable<Row>> apply =
+              inputData.apply(org.apache.beam.sdk.schemas.transforms.Group.globally());
+          inputData =
+              apply
+                  .apply(ParDo.of(uniquePartition(mockKeySchema, simulatedKeyValueSchema)))
+                  .setRowSchema(simulatedKeyValueSchema);
+        } else {
+          org.apache.beam.sdk.schemas.transforms.Group.ByFields<Row> myg =
+              org.apache.beam.sdk.schemas.transforms.Group.byFieldIds(af.partitionKeys);
+          inputData = inputData.apply("partitionBy", myg);
```

Comment:
In fact, if there is no order by but only partition by, combine per key will be the best API because backend can do combiner lift optimizations, which works as follow:

KV -> GroupByKey -> KV -> Combine
can be optimized as
KV -> local combine -> K CombinedV -> GroupByKey -> Combine all combined V.

So basically there could be a pre-combine before shuffle (GroupByKey), and after that each worker will only need to combine those pre-combined value (because of associativity rule).  This optimization will reduce lots of data through shuffle.

I think this can be left for a future work. Please log a JIRA to document this idea.

Reply:
I added a comment in the JIRA issue of the project [`issue`](https://issues.apache.org/jira/browse/BEAM-9198?focusedCommentId=17151728&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17151728)